### PR TITLE
Game state manager

### PR DESCRIPTION
**Pull request by jidaprano**
- Implemented game state manager
        - Current states include Main menu state, playing state, and paused state
- Added game manager to store state manager, esc menu controller, and other high-level management classes that we implement
- Changed input system in SettingsController and EscMenuController to serializable InputActions
- Most changed/new files in Scripts/GameManagement
- Scene testing occurred in was the MainMenu scene

### CODE REVIEW

File: Assets/Scripts/UIAndMenus/EscMenuController.cs [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
 using UnityEngine;
+using UnityEngine.InputSystem;
+using static UnityEngine.InputSystem.InputAction;
 
 public class EscMenuController : MonoBehaviour
 {
+    [SerializeField] private InputAction pauseAction;
+
     public GameObject pauseMenuPanel;
     public GameObject settingsMenuPanel;
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/EscMenuController.cs:14]: these should probably also be private SerializeFields (with properties to access them in other classes)

File: Assets/Scripts/UIAndMenus/EscMenuController.cs [hunk 2]
```diff
@@ -13,26 +18,37 @@ void Start()
     {
         pauseMenuPanel.SetActive(false);
         settingsMenuPanel.SetActive(false);
+        DontDestroyOnLoad(gameObject);
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/EscMenuController.cs:21]: image

Unity is throwing a warning saying that this line will not work--you should probably check this is doing what you want it to

File: Assets/Scripts/UIAndMenus/EscMenuController.cs [hunk 3]
```diff
@@ -13,26 +18,37 @@ void Start()
...
-        }
-        else
+            GameStateManager.Instance.SetState(new GameStateManagement.PauseState(this));
+        } else
         {
-            pauseMenuPanel.SetActive(true);
-            Time.timeScale = 0f;
+            GameStateManager.Instance.SetState(new PlayingState());
         }
     }
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/EscMenuController.cs:45]: Should add a check to determine whether in the main menu (currently, you can access the pause screen when in the main menu, which should not be possible)

File: Assets/Scripts/UIAndMenus/EscMenuController.cs [hunk 4]
```diff
@@ -47,6 +63,6 @@ public void CloseSettings()
 
     public void Quit()
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/EscMenuController.cs:64]: may want to change this to something other than Quit (like ReturnToMainMenu() maybe? unless you can come up with something better) to differentiate it from quitting the application
(probably not strictly needed, though--more of a stylistic choice)

File: Assets/Scripts/UIAndMenus/MainMenu.cs [hunk 5]
```diff
@@ -31,12 +32,5 @@ public void OnSettingsDoneButton()
     public void OnQuitButton()
     {
         Application.Quit();
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/MainMenu.cs:34]: Quitting the application should eventually be done in a different class to handle the logic for different platforms (ex. it looks like we have to do something different if we have a web version) (also, while I can't currently think of a reason why you would quit from anywhere other than the main menu, it is possible that we may)

Maybe a QuitHandler class belonging to the GameStateManager that can be called through the GameStateManager?

File: Assets/Scripts/UIAndMenus/SettingsController.cs [hunk 6]
```diff
@@ -35,9 +35,9 @@ public class SettingsController : MonoBehaviour
     // Start is called before the first frame update
```
Comment (smlendzian) [inline Assets/Scripts/UIAndMenus/SettingsController.cs:35]: This class is starting to get pretty long, so at some point it should probably be refactored into multiple classes (maybe separating the categories logic into a different class and/or putting the logic for specific features--like adjusting the resolution and changing the volume--into their own classes).

File: Assets/Scripts/UIAndMenus/MainMenu.cs (hunk 5 above)
Comment (jidaprano) [reply Assets/Scripts/UIAndMenus/MainMenu.cs:34]: Agreed, I'm gonna make a note of this in the save system work item and push it to there

File: Assets/Scripts/UIAndMenus/SettingsController.cs (hunk 6 above)
Comment (jidaprano) [reply Assets/Scripts/UIAndMenus/SettingsController.cs:35]: Agreed, I'm going to push this off a bit and make a new work item

**Pull request merged**